src/libbson/src/bson/bson-utf8.c

```diff
@@ -261,80 +372,108 @@ char *
+
+   size_t normal_chars_seen = 0u;
+
+   do {
+      const uint8_t current_byte = (uint8_t) utf8[normal_chars_seen];
+      if (!_is_special_char (current_byte)) {
+         // Normal character, no need to do anything besides iterate
+         // Copy rest of the string if we reach the end
+         normal_chars_seen++;
+         utf8_ulen--;
+         if (utf8_ulen == 0) {
+            _bson_string_append_ex (str, utf8, normal_chars_seen);
+            break;
+         }
+
+         continue;
+      }
+
+      // Reached a special character. Copy over all of normal characters
+      // we have passed so far
+      if (normal_chars_seen > 0) {
+         _bson_string_append_ex (str, utf8, normal_chars_seen);
+         utf8 += normal_chars_seen;
+         normal_chars_seen = 0;
+      }
 
       // Check if expected char length goes past end
-      _bson_utf8_get_sequence (utf8, &length_of_char, &mask);
-      if (utf8 > end - length_of_char) {
-         goto invalid_utf8;
+      // bson_utf8_get_char will crash without this check
+      {
+         uint8_t mask;
+         uint8_t length_of_char;
+
+         _bson_utf8_get_sequence (utf8, &length_of_char, &mask);
+         if (utf8 > end - length_of_char) {
+            goto invalid_utf8;
+         }
       }
 
-      c = bson_utf8_get_char (utf8);
-
-      switch (c) {
-      case '\\':
-      case '"':
-         bson_string_append_c (str, '\\');
-         bson_string_append_unichar (str, c);
-         break;
-      case '\b':
-         bson_string_append (str, "\\b");
-         break;
-      case '\f':
-         bson_string_append (str, "\\f");
-         break;
-      case '\n':
-         bson_string_append (str, "\\n");
-         break;
-      case '\r':
-         bson_string_append (str, "\\r");
-         break;
-      case '\t':
-         bson_string_append (str, "\\t");
-         break;
-      default:
-         if (c < ' ') {
-            bson_string_append_printf (str, "\\u%04x", (unsigned) c);
-         } else {
-            bson_string_append_unichar (str, c);
+      // Check for null character
+      // Null characters are only allowed if the length is provided
+      if (utf8[0] == '\0' || (utf8[0] == '\xc0' && utf8[1] == '\x80')) {
+         if (!length_provided) {
+            goto invalid_utf8;
          }
-         break;
+
+         bson_string_append (str, "\\u0000");
+         utf8_ulen -= *utf8 ? 2u : 1u;
+         utf8 += *utf8 ? 2 : 1;
+         continue;
       }
 
-      if (c) {
-         utf8 = bson_utf8_next_char (utf8);
-      } else {
-         if (length_provided) {
-            // Check if current character is null character,
-            // if so skip past it as we've already added '\\u0000' above
-            if (*utf8 == '\0' || (*utf8 == '\xc0' && *(utf8 + 1) == '\x80')) {
-               utf8 += (*utf8 == '\0') ? 1 : 2;
-            } else {
-               goto invalid_utf8;
-            }
-         } else {
+      // Highest ASCII character
```

Comment:
```suggestion
      // Multi-byte UTF-8 sequence
```

I expect this block only applies to multi-byte UTF-8 sequences.